### ISSUE TITLE
Support localization of headers (translatability)

### ISSUE THREAD

**Issue by AILazerka** [kolkov/angular-editor#304]
It would be great to achieve customizable translatability of the `title`s for a use in products which are localized for different regions to exclude need in manual hacks.

Thanks in advance!

**Comment by kolkov** [comment]
Yes, we discussed such a possibility, but so far there is no good solution, how it can be get it in the library.  I understand that the Angular team will work on adding the library localization possibility in the new Ivy library format.

**Comment by wandu-ar** [comment]
May be can solve it with an input attribute passing a translation object that next we can translate with angular i18n tools ...

For example:

`<angular-editor formControlName="htmlContent" [config]="editorConfig" [locales]="{bold: 'Bold', italic; 'Italic' ... }" i18n-locales></angular-editor>`

Thanks!

**Comment by AILazerka** [comment]
Hello @kolkov! I understand. Right now in a project we are in need of using a workaround for making headers appear in German. And personally I can say it looks not that good. From the answer above i18n-locales can be used. We can also use `ngx-translate` I believe with some fallback default translation. We can reserve JSON section for it like:

```
{
  ...
  "AngularEditor": {
    ...
  },
  ...
}
```
or even require separate json files for translations.

And when importing the library we can provide translation loader in the config for the AngularEditor and this loader will be pointing to either general file with the structure above or a separate file with the translations which can be placed in the projects under assets as `angular-editor-en.json`, `angular-editor-de.json` and other languages or any completely different name. What do you think?

**Comment by SiriRavoori** [comment]
Can you update on this implementation of translations(for tool bar) ?

**Comment by kolkov** [comment]
I was googled from any news about translation, but I'm not tryed to use this ability yet.
https://github.com/angular/angular/issues/29536

**Comment by SiriRavoori** [comment]
Hi,

Thank you for your reply.
I have managed translations using JQuery.

Regards,
Sireesha Naraharisetti

**Comment by mannyluccio** [comment]
@SiriRavoori Any advice on how you handled it with jquery?

**Comment by ThamiresLC** [comment]
Gostaria de colocar o meu projeto em Português-BR alguém tem alguma solução?

**Comment by cortaza123** [comment]
¿Hay alguna forma de traducir los nombres de los botones a otro idioma?, como lo manejaste con JQuery, saludos.

**Comment by kolkov** [comment]
Thank you for this feature request!

## Feature: Toolbar Localization/Translation

This is a **valid and frequently requested feature** for multi-language applications.

### Current Limitation

Toolbar buttons have **hardcoded English titles**:
- "Bold" (title attribute)
- "Italic"
- "Underline"
- "Insert Link"
- etc.

### Proposed Solutions

Several approaches have been discussed:

#### Option 1: Translation Object Input

```typescript
@Component({
  template: `
    <angular-editor
      [config]="config"
      [translations]="translations">
    </angular-editor>
  `
})
export class EditorComponent {
  translations = {
    bold: 'Fett',
    italic: 'Kursiv',
    underline: 'Unterstrichen',
    insertLink: 'Link einfügen',
    // ... all toolbar items
  };
}
```

#### Option 2: Use ngx-translate

```typescript
import { TranslateService } from '@ngx-translate/core';

config: AngularEditorConfig = {
  translations: this.translate.instant('AngularEditor')
};

// In translation file (de.json):
{
  "AngularEditor": {
    "bold": "Fett",
    "italic": "Kursiv",
    "underline": "Unterstrichen"
  }
}
```

#### Option 3: Built-in i18n Support

```typescript
// Library provides translations
import { ANGULAR_EDITOR_TRANSLATIONS_DE } from '@kolkov/angular-editor/i18n';

config: AngularEditorConfig = {
  translations: ANGULAR_EDITOR_TRANSLATIONS_DE
};
```

#### Option 4: Angular i18n (Future)

Once Angular library localization is stable:

```html
<angular-editor 
  [config]="config"
  i18n-title>
</angular-editor>
```

### Implementation Approach

**Files to modify:**
- `config.ts` - Add `translations` interface
- `ae-toolbar.component.ts` - Use translations instead of hardcoded strings
- `ae-button.component.ts` - Accept translated title

**Interface:**

```typescript
export interface AngularEditorTranslations {
  // Font/Text
  bold?: string;
  italic?: string;
  underline?: string;
  strikeThrough?: string;
  subscript?: string;
  superscript?: string;
  
  // Alignment
  justifyLeft?: string;
  justifyCenter?: string;
  justifyRight?: string;
  justifyFull?: string;
  
  // Lists
  orderedList?: string;
  unorderedList?: string;
  indent?: string;
  outdent?: string;
  
  // Insert
  insertLink?: string;
  insertImage?: string;
  insertVideo?: string;
  insertHorizontalRule?: string;
  
  // Format
  textColor?: string;
  backgroundColor?: string;
  fontName?: string;
  fontSize?: string;
  heading?: string;
  
  // Other
  removeFormat?: string;
  toggleEditorMode?: string;
  undo?: string;
  redo?: string;
}

export interface AngularEditorConfig {
  // ... existing config
  translations?: AngularEditorTranslations;
}

// Default English translations
export const ANGULAR_EDITOR_TRANSLATIONS_EN: AngularEditorTranslations = {
  bold: 'Bold',
  italic: 'Italic',
  underline: 'Underline',
  // ...
};
```

**Usage in component:**

```typescript
export class AeToolbarComponent {
  @Input() translations?: AngularEditorTranslations;
  
  get boldTitle(): string {
    return this.translations?.bold || 'Bold';
  }
  
  get italicTitle(): string {
    return this.translations?.italic || 'Italic';
  }
}
```

### Current Workaround (jQuery/DOM Manipulation)

As mentioned in comments, some users use jQuery:

```typescript
ngAfterViewInit() {
  // Change toolbar titles after rendering
  $('.angular-editor-toolbar button[title="Bold"]').attr('title', 'Fett');
  $('.angular-editor-toolbar button[title="Italic"]').attr('title', 'Kursiv');
  // ... etc
}
```

**Better workaround without jQuery:**

```typescript
ngAfterViewInit() {
  const toolbar = document.querySelector('.angular-editor-toolbar');
  
  const translations = {
    'Bold': 'Fett',
    'Italic': 'Kursiv',
    'Underline': 'Unterstrichen',
    'Insert Link': 'Link einfügen'
  };
  
  toolbar?.querySelectorAll('button[title]').forEach((button: HTMLElement) => {
    const currentTitle = button.getAttribute('title');
    if (currentTitle && translations[currentTitle]) {
      button.setAttribute('title', translations[currentTitle]);
    }
  });
}
```

### Recommended Implementation

**For the library:**

1. Add `translations` config option
2. Provide default English translations
3. Ship common language packs (optional):
   - `@kolkov/angular-editor/i18n/de`
   - `@kolkov/angular-editor/i18n/fr`
   - `@kolkov/angular-editor/i18n/es`
   - etc.

4. Allow custom translations via config

**Example:**

```typescript
import { ANGULAR_EDITOR_TRANSLATIONS_DE } from '@kolkov/angular-editor/i18n';

config: AngularEditorConfig = {
  // Use built-in German
  translations: ANGULAR_EDITOR_TRANSLATIONS_DE,
  
  // Or provide custom
  translations: {
    bold: 'مُسَوَّد', // Arabic
    italic: 'مائل',
    // ...
  }
};
```

### Impact

**HIGH** - Essential for:
- International applications
- Multi-language products
- Accessibility (screen readers in different languages)
- User experience in non-English regions

### Complexity

**MEDIUM** - Requires:
- Interface definition
- Refactoring hardcoded strings
- Default translations
- Optional: Additional language packs

### Related

Angular team is working on library localization:
- https://github.com/angular/angular/issues/29536

Once stable, the library could migrate to Angular's built-in i18n.

## Community Contribution Opportunity

This would be an **excellent community contribution**! Steps:

1. Define `AngularEditorTranslations` interface
2. Refactor hardcoded titles to use translations
3. Create language packs for common languages
4. Document usage in README

**Would anyone like to contribute this feature?** Please comment if you're interested!

**Keeping this OPEN** as a high-priority feature request with strong community interest.
